Cloud Auth plugin & Google Pub/Sub destination

alltilla opened the pull request:
#### `cloud-auth()`: Added a new plugin for drivers, which implements different cloud related authentications.

Currently the only supported authentication is [GCP's Service Account](https://cloud.google.com/iam/docs/service-account-overview) for the `http()` destination.

Example config:
```
http(
  cloud-auth(
    gcp(
      service-account(
        key("/path/to/service-account-key.json")
        audience("https://pubsub.googleapis.com/google.pubsub.v1.Publisher")
      )
    )
  )
);
```

---

#### Sending messages to Google Pub/Sub

The `google-pubsub()` destination feeds Google Pub/Sub via the [HTTP REST API](https://cloud.google.com/pubsub/docs/reference/rest/v1/projects.topics/publish).

Example config:
```
google-pubsub(
  project("syslog-ng-project")
  topic("syslog-ng-topic")
  auth(
    service-account(
      key("/path/to/service-account-key.json")
    )
  )
);
```

See the [Google Pub/Sub documentation](https://cloud.google.com/pubsub/docs/building-pubsub-messaging-system) to learn more about configuring a service account.


Signed-off-by: Attila Szakacs <attila.szakacs@axoflow.com>

furiel commented on modules/cloud-auth/cloud-auth.hpp:
```diff
@@ -0,0 +1,51 @@
+  void (*free_fn)(CloudAuthenticator *s);
+};
+
+#endif
```
nit: newline

kira-syslogng commented:
Build FAILURE

furiel commented on modules/cloud-auth/google-auth.hpp:
```diff
@@ -0,0 +1,59 @@
+  std::string private_key_id;
+};
+
+}
```
I suppose the indentation is intentional here. I just leave a note here to confirm.

alltilla replied:
Yes, we decided to format the namespaces like this.

furiel commented on modules/cloud-auth/google-auth.cpp:
```diff
@@ -0,0 +1,188 @@
+  g_string_append(buffer, "Authorization: Bearer ");
+
+  std::chrono::system_clock::time_point issued_at = jwt::date::clock::now();
+  std::chrono::system_clock::time_point expires_at = issued_at + std::chrono::seconds{3600};
```
Is this 3600 configurable? Do we plan to configure? Could you give a name for it?

alltilla replied:
Yes, good idea. Fixed! :)

kira-syslogng commented:
Build FAILURE

alltilla commented on modules/cloud-auth/CMakeLists.txt:
```diff
@@ -0,0 +1,43 @@
+  message(FATAL_ERROR "C++ support is mandatory when the Cloud Auth module is enabled.")
+endif()
+
+module_switch(ENABLE_CLOUD_AUTH "Enable Cloud Auth" ENABLE_CPP)
```
I copied this from other modules, and it seems to work, yes.

HofiOne replied:
the idea behind module_switch is that it will define e.g ENABLE_CLOUD_AUTH if it is not already defined that way you can control options like this from the cmake command line as well e.g. -DENABLE_CLOUD_AUTH=OFF

kira-syslogng commented:
Build FAILURE

MrAnno reviewed (commented):
LGTM so far 👍

alltilla commented:
I have fixed the review comments, modified a bit on the configurability and added news entries. This is ready for review again. @MrAnno @furiel Could you kindly take a look again? Thanks!

alltilla commented:
Sry, pushed again.
